src/notification/services/email.service.ts

```diff
@@ -0,0 +1,117 @@
+      const rateRepository = new RateService(new RatesRepository());
+      const currentRate = await rateRepository.getLatest(Currency.USD);
+      if (!currentRate) {
+        logger.error('No currency rate data found');
+        throw new Error('No currency rate data found');
+      }
+      const subscriptionService = new SubscriptionsService(new SubscriptionsRepository());
+      const requestsLimit = config.api.emailServer.rateLimit;
+
+      logger.info(`[Scheduled job] [${new Date()}] going to send emails to subsribed users`);
+      let subcriptions = await subscriptionService.getAll({ limit: requestsLimit });
+      let hasMore = !!subcriptions.length;
+      const failedToSendEmailErrors: string[] = [];
+      while (hasMore) {
+        const sendEmailPromises = subcriptions.map((subscription) => {
+          return this.sendEmail(this.buildCurrencyRateEmailPayload({ to: subscription.email, currencyRate: currentRate.value }));
+        });
+        const results = await Promise.allSettled(sendEmailPromises);
+        const failedRequests = results.filter((result) => {
+          return result.status === 'rejected';
+        });
+        failedRequests.forEach((result) => {
+          const error = ((result as PromiseRejectedResult) || undefined)?.reason;
+          failedToSendEmailErrors.push(`Error while sending email: ${error?.message}. Details: ${JSON.stringify(error)}`);
+        });
+
+        const createdAfter = subcriptions[subcriptions.length - 1].createdAt;
+        subcriptions = await subscriptionService.getAll({ limit: requestsLimit, createdAfter });
+        hasMore = !!subcriptions.length;
+      }
+
+      if (failedToSendEmailErrors.length) {
+        logger.error(`Errors sending currency exchange rate: ${JSON.stringify(failedToSendEmailErrors)}`);
+      }
+    } catch (error) {
+      logger.error(`Got an error when sending daily rate email notifications ${JSON.stringify(error)}`);
+    }
+  }
+
+  public async sendEmail(payload: EmailPayload) {
+    return this.emailSender.sendMail(payload);
+  }
+
+  private buildCurrencyRateEmailPayload(params: CurrencyRateEmailPayload): EmailPayload {
+    return {
+      from: config.api.emailServer.user,
+      subject: 'Currency Rate USD to UAH',
+      message: `${params.currencyRate}`,
+      to: params.to,
+    };
+  }
+
+  private async saveCurrencyData(data: { currencyRate: number }) {
+    const rateRepository = new RatesRepository();
+    await rateRepository.create(data.currencyRate, Currency.USD);
+  }
+
+  private async systemEventHandler(event: any): Promise<void> {
+    if (!event?.message?.value) {
+      console.error('Empty message from Message Broker');
+      return;
+    }
+    const eventPayload = JSON.parse(event.message.value.toString()) as SystemEvent;
+    console.log('Received event:', eventPayload);
+    switch (eventPayload.eventType) {
+      case SystemEventType.CurrencyRateUpdated:
+        await this.saveCurrencyData(eventPayload.data);
```

Comment:
In case the rate repository is embedded in email sender you don't need to write a separate function to save the rate to the database